tools/testbench/common_test_ipc3.c

```diff
@@ -267,38 +217,212 @@ int tb_pipeline_params(struct testbench_prm *tp, struct ipc *ipc, struct pipelin
+	struct list_item *temp;
+	struct ipc_comp_dev *icd = NULL;
+	int err;
+
+	/* remove the components for this pipeline */
+	list_for_item_safe(clist, temp, &sof_get()->ipc->comp_list) {
+		icd = container_of(clist, struct ipc_comp_dev, list);
+
+		switch (icd->type) {
+		case COMP_TYPE_COMPONENT:
+			if (icd->cd->pipeline->pipeline_id != pipeline_id)
+				break;
+			err = ipc_comp_free(sof_get()->ipc, icd->id);
+			if (err)
+				fprintf(stderr, "failed to free comp %d\n", icd->id);
+			break;
+		case COMP_TYPE_BUFFER:
+			if (buffer_pipeline_id(icd->cb) != pipeline_id)
+				break;
+			err = ipc_buffer_free(sof_get()->ipc, icd->id);
+			if (err)
+				fprintf(stderr, "failed to free buffer %d\n", icd->id);
+			break;
+		case COMP_TYPE_PIPELINE:
+			if (icd->pipeline->pipeline_id != pipeline_id)
+				break;
+			err = ipc_pipeline_free(sof_get()->ipc, icd->id);
+			if (err)
+				fprintf(stderr, "failed to free pipeline %d\n", icd->id);
+			break;
+		default:
+			fprintf(stderr, "Unknown icd->type %d\n", icd->type);
+		}
+	}
+}
+
+int tb_free_all_pipelines(struct testbench_prm *tp)
+{
+	int i;
+
+	for (i = 0; i < tp->pipeline_num; i++)
+		test_pipeline_free_comps(tp->pipelines[i]);
+
+	return 0;
+}
+
+void tb_free_topology(struct testbench_prm *tp)
 {
-	if (host_trace_level >= LOG_LEVEL_DEBUG)
-		printf("debug: %s", message);
 }
 
-/* enable trace in testbench */
-void tb_enable_trace(unsigned int log_level)
+static int test_pipeline_params(struct testbench_prm *tp)
 {
-	host_trace_level = log_level;
-	if (host_trace_level)
-		debug_print("trace print enabled\n");
-	else
-		debug_print("trace print disabled\n");
+	struct ipc_comp_dev *pcm_dev;
+	struct pipeline *p;
+	struct ipc *ipc = sof_get()->ipc;
+	int ret = 0;
+	int i;
+
+	/* Run pipeline until EOF from fileread */
+
+	for (i = 0; i < tp->pipeline_num; i++) {
+		pcm_dev = ipc_get_ppl_src_comp(ipc, tp->pipelines[i]);
+		if (!pcm_dev) {
+			fprintf(stderr, "error: pipeline %d has no source component\n",
+				tp->pipelines[i]);
+			return -EINVAL;
+		}
+
+		/* set up pipeline params */
+		p = pcm_dev->cd->pipeline;
+
+		/* input and output sample rate */
+		if (!tp->fs_in)
+			tp->fs_in = p->period * p->frames_per_sched;
+
+		if (!tp->fs_out)
+			tp->fs_out = p->period * p->frames_per_sched;
+
+		ret = tb_pipeline_params(tp, ipc, p);
+		if (ret < 0) {
+			fprintf(stderr, "error: pipeline params failed: %s\n",
+				strerror(ret));
+			return ret;
+		}
+	}
+
+	return 0;
 }
 
-void tb_gettime(struct timespec *td)
+static void tb_test_pipeline_set_test_limits(int pipeline_id, int max_copies,
+					     int max_samples)
 {
-#if !defined __XCC__
-	clock_gettime(CLOCK_MONOTONIC, td);
-#else
-	td->tv_nsec = 0;
-	td->tv_sec = 0;
-#endif
+	struct list_item *clist;
+	struct list_item *temp;
+	struct ipc_comp_dev *icd = NULL;
+	struct comp_dev *cd;
+	struct dai_data *dd;
+	struct file_comp_data *fcd;
+
+	/* set the test limits for this pipeline */
+	list_for_item_safe(clist, temp, &sof_get()->ipc->comp_list) {
```

Comment:
here no list items are removed or moved or added, so it doesn't need to be "safe?"